purchase_discount/models/purchase_order.py

```diff
@@ -47,18 +49,28 @@ class PurchaseOrderLine(models.Model):
 
     @api.depends('discount')
     def _compute_amount(self):
+        """ Inject the product price with proper rounding in the context from
+        which account.tax::compute_all() is able to retrieve it. The alternate
+        context is patched onto self because it can be a NewId passed in the
+        onchange the env of which does not support `with_context`. """
         for line in self:
-            price_unit = False
+            orig_context = None
             # This is always executed for allowing other modules to use this
             # with different conditions than discount != 0
-            price = line._get_discounted_price_unit()
-            if price != line.price_unit:
-                # Only change value if it's different
-                price_unit = line.price_unit
-                line.price_unit = price
+            discounted_price_unit = line._get_discounted_price_unit()
+            if discounted_price_unit != line.price_unit:
+                precision = line.order_id.currency_id.decimal_places
+                company = line.company_id or self.env.user.company_id
+                if company.tax_calculation_rounding_method == 'round_globally':
+                    precision += 5
+                orig_context = self.env.context
+                price = round(
+                    line.product_qty * discounted_price_unit, precision)
+                self.env.context = frozendict(
```

Comment:
Changing context this way can be a little hackish and maybe affect cache mechanism. Are you sure there's no side effect?

Reply:
I have not noticed any side effect (I'm using it elsewhere too) but you may ask @Yajo https://github.com/odoo/odoo/issues/7472#issuecomment-247058243

Reply:
Indeed this is a hack that should be avoided if possible. AFAIK the only use case where it is justified is for onchanges.

Please change it for:

```python
self.with_context(base_values=(price, price, price))._compute_amount()
```

which is the correct way of altering context.

Reply:
The problem with that, @Yajo, is that compute fields depend on that context, so we won't get the achieved effect. That's why we initially replace the field value instead of doing this.

Reply:
@yajo this is an onchange

Reply:
OK, I thought that computed fields did not have this problem. It's OK then.